src/addons/meta/runtime.c

```diff
@@ -0,0 +1,328 @@
+static void flecs_meta_rtt_copy(void *dst_ptr, const void *src_ptr, int32_t count, const ecs_type_info_t *type_info) {
+  ecs_meta_rtt_ctx_t *rtt_ctx = (ecs_meta_rtt_ctx_t *) type_info->hooks.lifecycle_ctx;
+  ecs_assert(rtt_ctx != NULL, ECS_INTERNAL_ERROR, NULL);
+
+  int cb_count = ecs_vec_count(&rtt_ctx->copys);
+  for (int j = 0; j < count; j++) {
+    ecs_size_t elem_offset = type_info->size * j;
+    void *elem_dst_ptr = ECS_OFFSET(dst_ptr, elem_offset);
+    const void *elem_src_ptr = ECS_OFFSET(src_ptr, elem_offset);
+    for (int i = 0; i < cb_count; i++) {
+      ecs_meta_rtt_call_data_t *copy_data = ecs_vec_get_t(&rtt_ctx->copys, ecs_meta_rtt_call_data_t, i);
+      copy_data->hook.copy(ECS_OFFSET(elem_dst_ptr, copy_data->offset), ECS_OFFSET(elem_src_ptr, copy_data->offset),
+                           copy_data->count, copy_data->type_info);
+    }
+  }
+}
+
+//
+static void flecs_meta_rtt_default_move(void *dst_ptr, void *src_ptr, int32_t count, const ecs_type_info_t *type_info) {
+  (void) type_info;  // silence unused warning
+  ecs_os_memcpy(dst_ptr, src_ptr, count);
+}
+
+static void flecs_meta_rtt_default_copy(void *dst_ptr, const void *src_ptr, int32_t count,
+                                        const ecs_type_info_t *type_info) {
+  (void) type_info;  // silence unused warning
+  ecs_os_memcpy(dst_ptr, src_ptr, count);
+}
+
+static void flecs_meta_rtt_free_lifecycle_ctx(void *ctx) {
+  if (!ctx)
+    return;
+
+  ecs_meta_rtt_ctx_t *lifecycle_ctx = (ecs_meta_rtt_ctx_t *) ctx;
+
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->ctors, ecs_meta_rtt_call_data_t);
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->dtors, ecs_meta_rtt_call_data_t);
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->moves, ecs_meta_rtt_call_data_t);
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->copys, ecs_meta_rtt_call_data_t);
+
+  ecs_os_free(ctx);
+}
+
+static ecs_meta_rtt_ctx_t *flecs_meta_rtt_configure_hooks(ecs_world_t *world, const ecs_type_info_t *ti, bool ctor,
+                                                          bool dtor, bool move, bool copy) {
+  ecs_type_hooks_t hooks = ti->hooks;
+  flecs_meta_rtt_free_lifecycle_ctx(hooks.lifecycle_ctx);
+  ecs_meta_rtt_ctx_t *rtt_ctx = NULL;
+  if (ctor || dtor || move || copy) {
+    rtt_ctx = ecs_os_malloc_t(ecs_meta_rtt_ctx_t);
+    ecs_vec_init_t(NULL, &rtt_ctx->ctors, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->dtors, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->moves, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->copys, ecs_meta_rtt_call_data_t, 0);
+    hooks.lifecycle_ctx = rtt_ctx;
+    hooks.lifecycle_ctx_free = flecs_meta_rtt_free_lifecycle_ctx;
+
+    if (ctor) {
+      hooks.ctor = flecs_meta_rtt_ctor;
+    }
+    if (dtor) {
+      hooks.dtor = flecs_meta_rtt_dtor;
+    }
+    if (move) {
+      hooks.move = flecs_meta_rtt_move;
+    }
+    if (copy) {
+      hooks.copy = flecs_meta_rtt_copy;
+    }
+  } else {
+    hooks.lifecycle_ctx = NULL;
```

Comment:
I have to set it to `NULL` because this could be the case that the hooks are being modified from requiring hooks to not requiring hooks. In this case I can't leave `lifecycle_ctx` or `lifecycle_ctx_free` function pointing to something, since there is no more `lifecycle_ctx`. That would leave the hooks in an inconsistent state.

That said, I modified it a bit to also clear `lifecycle_ctx_free`.

Reply:
> because this could be the case that the hooks are being modified from requiring hooks to not requiring hooks

What is a scenario in which that could happen?

If it is possible that `lifecycle_ctx` is already set to something, wouldn't setting the value to `NULL` without freeing it result in a memory leak?

Reply:
I don't know what scenario, but does not hurt to be clean, set the pointer to NULL after freeing, otherwise an attempt to free would happen again on world fini.

> If it is possible that lifecycle_ctx is already set to something, wouldn't setting the value to NULL without freeing it result in a memory leak?

It is freed above. Note that every time a member is added to a struct, the context is freed and then the call list is regenerated from scratch, like TypeSerializer.

> Can you add a test that fails when this value is not set to NULL?

I'll look into this, but I belive such test is together already with the tests of binding_ctx and ctx.

Reply:
> Note that every time a member is added to a struct, the context is freed and then the call list is regenerated

Gotcha, that answers my question :)